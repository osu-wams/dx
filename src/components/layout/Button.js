import styled from 'styled-components';
import PropTypes from 'prop-types';

// Outputs the padding size in pixel for the component
function btnSize(size, theme) {
  if (size === 'small') {
    return `
      padding: ${theme.spacing.unit / 4}px ${theme.spacing.unit / 2}px;
      font-size: ${theme.fontSize.small};
    `;
  }
  if (size === 'large') {
    return `
      padding: ${theme.spacing.unit * 2}px ${theme.spacing.unit * 4}px;
      font-size: ${theme.fontSize.large};
    `;
  }
  return `
      padding: ${theme.spacing.unit}px ${theme.spacing.unit * 2}px;
      font-size: ${theme.fontSize.normal};
    `;
}

const Button = styled.button`
  background-color: ${({ outline, theme, variant }) =>
    outline ? 'transparent' : theme[variant].bg};
  color: ${({ outline, theme, variant }) => (outline ? theme[variant].bg : theme[variant].fg)};
  text-decoration: none;
  border-radius: 0.4rem;
  border: 1px solid ${({ theme, variant }) => theme[variant].bg};
  ${({ theme, size }) => btnSize(size, theme)};
  & + & {
    margin-left: ${props => props.theme.spacing.unit}px;
  }
`;

Button.propTypes = {
  variant: PropTypes.oneOf(['mist', 'primary', 'secondary', 'academic', 'finance']),
  size: PropTypes.oneOf(['small', 'normal', 'large'])
};

Button.defaultProps = {
  variant: 'mist',
  size: 'normal',
  outline: false
};

export default Button;
