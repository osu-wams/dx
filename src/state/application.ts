import { atom } from 'recoil';
import { User } from '@osu-wams/hooks';
import { Types } from '@osu-wams/lib';
import { defaultTheme } from 'src/theme/themes';

export const userState = atom<Types.UserState>({
  key: 'userState',
  default: { data: User.INITIAL_USER, loading: true, error: false },
});

export const themeState = atom<string>({
  key: 'themeState',
  default: defaultTheme,
});

export const infoButtonState = atom<{ content: string; id: string; title: string }[]>({
  key: 'infoButtonState',
  default: [],
});

export const plannerItemState = atom<{
  data: Types.PlannerItem[];
  isLoading: boolean;
  error: Error | null;
}>({
  key: 'plannerItemState',
  default: { data: [], isLoading: false, error: null },
});
