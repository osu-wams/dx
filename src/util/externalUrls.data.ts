const Url = {
  myDegrees: {
    main: 'https://mydegrees.oregonstate.edu/DashboardServlet'
  },
  canvas: {
    main: 'https://oregonstate.instructure.com',
    test: 'https://oregonstate.test.instructure.com'
  },
  campusMap: {
    main: 'https://map.oregonstate.edu/',
    building: 'https://map.oregonstate.edu/?building='
  },
  events: {
    academicCalendar: 'https://registrar.oregonstate.edu/osu-academic-calendar'
  },
  banner: {
    financialTransactions:
      'https://xe.ucsadm.oregonstate.edu:9990/ssomanager/c/SSB?pkg=OSU_ARWEB.P_ViewBillDate'
  },
  myosu: {
    main: 'https://myosu.oregonstate.edu/'
  },
  bill: {
    main: 'http://mybill.oregonstate.edu'
  },
  support: {
    main: 'https://oregonstate.qualtrics.com/jfe/form/SV_73vMvyQZBQx8aaN?type=support'
  },
  feedback: {
    main: 'https://oregonstate.qualtrics.com/jfe/form/SV_73vMvyQZBQx8aaN?type=feedback'
  },
  registrar: {
    academicStanding: 'https://registrar.oregonstate.edu/grades-honor-roll-academic-standing'
  },
  empcenter: {
    main: 'https://osu.workforcehosting.com/'
  },
  evals: {
    main: 'https://myosu.oregonstate.edu/web/myosu/employee?gpi=10190'
  },
  itSystemStatus: {
    main: 'https://status.is.oregonstate.edu'
  }
};

export default Url;
