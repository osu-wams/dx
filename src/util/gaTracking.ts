import ReactGA from 'react-ga';

/**
 * Event - Add custom tracking event.
 * @param {enum} category
 * @param {string} action
 * @param {string} label
 */
export const Event = (category: IComponents, action: string, label?: string) => {
  ReactGA.event({
    category,
    action,
    label,
  });
};

// To make it easier to filter, our components will be the categories in Google Analytics
export type IComponents =
  | 'footer'
  | 'beta'
  | 'beta-info'
  | 'header'
  | 'student-navigation-main'
  | 'employee-navigation-main'
  | 'student-mobile-nav'
  | 'footer-menu-nav'
  | 'employee-tools'
  | 'schedule-card'
  | 'resources-card'
  | 'favorite-resources-card'
  | 'favorite-resource'
  | 'trending-resources-card'
  | 'trending-resource'
  | 'anouncements'
  | 'events'
  | 'courses'
  | 'course'
  | 'planner-items'
  | 'meal-plans'
  | 'dx-event'
  | 'calendar-event'
  | 'academic-calendar'
  | 'program-of-study'
  | 'financial-transactions'
  | 'academic-overview'
  | 'account-balance'
  | 'resource'
  | '404'
  | 'resource-search'
  | 'resource-search-failed'
  | 'resource-category'
  | 'past-courses-search'
  | 'info-button'
  | 'canvas-auth'
  | 'badge'
  | 'it-system-status'
  | 'it-system-status-sticky'
  | 'training';
