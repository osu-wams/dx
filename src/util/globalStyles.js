import { createGlobalStyle } from 'styled-components';
import { fontSize } from '@osu-wams/theme';

const GlobalStyle = createGlobalStyle`
  html {
    box-sizing: border-box;
    font-size: 62.5%;
    font-family: 'Open Sans', sans-serif;
    *, *:before, *:after {
      box-sizing: inherit;
    }
  }
  body {
    padding: 0;
    margin: 0;
    font-size: ${fontSize[16]};
    line-height: 1.5;
    color: ${({ theme }) => theme.body.color};
  }
  body:not(.user-is-tabbing) button:focus,
  body:not(.user-is-tabbing) input:focus,
  body:not(.user-is-tabbing) select:focus,
  body:not(.user-is-tabbing) textarea:focus {
    outline: none;
  }
  #root {
    background-color: ${({ theme }) => theme.body.background};
    min-height: 100vh;
    display: flex;
    flex-direction: column;
  }
  input, select, textarea, button {
    font-size: inherit;
  }
  strong {
    font-weight: 600;
  }
  /* Router helper */
  .router-styles {
    display: flex;
    height: 100%;
    flex-direction: column;
  }
  [data-reach-dialog-overlay] {
    z-index: 100;
  }
  @media screen and (max-width: 767px) {
    [data-reach-dialog-overlay] {
      background: ${({ theme }) => theme.ui.myDialog.background} !important;
    }
  }
  @media screen and (min-width: 768px) {
    [data-reach-menu] {
      z-index: 10;
    }
  }
  .gsc-selected-option-container {
    /* default is 90% and causes a weird wrapping effect */
    max-width: 100% !important;
  }
`;

export default GlobalStyle;
