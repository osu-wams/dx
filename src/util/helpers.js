import format from 'date-fns/format';

// Format any WORDS SENTENCE/word sentence/wOrD sEnTenCe to Title Case: Word Sentence
export const titleCase = phrase =>
  phrase
    .toLowerCase()
    .split(' ')
    .map(word => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');

// Format API time values to human friendly 12 hour am/pm format
export function formatTime(hours) {
  let result = 'Unexpected time format';

  // Expected time format from API
  const expectedFormat = /^\s*([01]?\d|2[0-3]):?([0-5]\d):?([0-5]\d)\s*$/;
  const matches = hours.match(expectedFormat);

  // Expected format matches user input, so we transform it
  if (matches) {
    const hourEnd = hours.indexOf(':');
    const H = hours.substr(0, hourEnd);
    const h = H % 12 || 12;
    const suffix = H < 12 ? 'am' : 'pm';
    result = h + hours.substr(hourEnd, 3) + suffix;
  }
  return result;
}

/* Format our day and time into preferred format
using the npm package date-fns. We change number dates to:
"December 09, 2018" format. */
export function formatDate(date) {
  return format(date, 'MMMM DD, YYYY');
}
