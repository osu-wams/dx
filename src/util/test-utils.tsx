import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render as testingLibraryRender } from '@testing-library/react';
import { Context as ResponsiveContext } from 'react-responsive';
import { themesLookup, defaultTheme } from '../theme/themes';
import { User } from '@osu-wams/lib';
import { mobile, desktop } from 'src/util/useMediaQuery';
import { RecoilRoot } from 'recoil';
import { userState } from 'src/state/application';
import { rest } from 'msw';
import { server } from 'src/mocks/server';

// Helper method to change the mock responses by MSW
export const alterMock = (api: string, mock: any) => {
  server.use(
    rest.get(api, async (req, res, ctx) => {
      return res(ctx.json(mock));
    })
  );
};

const { mockUser } = User;
export const authUserAudienceOverride = mockUser.userAudienceOverride;

export const mockGradUser = {
  ...mockUser.userGraduate,
  refreshFavorites: jest.fn(),
};

export const authUserClassification = mockUser.userClassification;

export const mockEmployeeUser = {
  ...mockUser.userEmployee,
  refreshFavorites: jest.fn(),
};

export const mockStudentEmployeeUser = {
  ...mockUser.userStudentEmployee,
  refreshFavorites: jest.fn(),
};

// The default undegraduate user with canvas opted in
export const authUser = {
  ...mockUser.user,
  refreshFavorites: jest.fn(),
};

// The default undegraduate user with canvas opted in
export const mockAdminUser = {
  ...mockUser.userAdmin,
  refreshFavorites: jest.fn(),
};

const renderWithUserContext = (
  ui,
  { user = authUser, initialStates = new Array(), ...options } = {}
) => {
  const Wrapper = (props) => {
    return (
      <RecoilRoot
        initializeState={(snap) => {
          snap.set(userState, user);
          initialStates.forEach((s: { state: any; value: any }) => snap.set(s.state, s.value));
        }}
      >
        <ThemeProvider theme={themesLookup[defaultTheme]} {...props} />
      </RecoilRoot>
    );
  };
  return testingLibraryRender(ui, { wrapper: Wrapper, ...options });
};

const renderWithAppContext = (ui, { initialStates = new Array(), ...options } = {}) => {
  const Wrapper = (props) => {
    return (
      <RecoilRoot
        initializeState={(snap) => {
          snap.set(userState, authUser);
          initialStates.forEach((s: { state: any; value: any }) => snap.set(s.state, s.value));
        }}
      >
        <ThemeProvider theme={themesLookup[defaultTheme]} {...props} />
      </RecoilRoot>
    );
  };
  return testingLibraryRender(ui, { wrapper: Wrapper, ...options });
};

const renderWithAllContexts = (
  ui,
  { user = authUser, isDesktop = false, initialStates = new Array(), ...options } = {}
) => {
  const Wrapper = (props) => {
    return (
      <RecoilRoot
        initializeState={(snap) => {
          snap.set(userState, user);
          initialStates.forEach((s: { state: any; value: any }) => snap.set(s.state, s.value));
        }}
      >
        <ThemeProvider theme={themesLookup[defaultTheme]}>
          <ResponsiveContext.Provider value={{ width: isDesktop ? desktop : mobile }} {...props} />
        </ThemeProvider>
      </RecoilRoot>
    );
  };
  return testingLibraryRender(ui, { wrapper: Wrapper, ...options });
};

const render = renderWithAllContexts;
// Pass a different user
// const { getByTestId } = renderWithUserContext(<Dashboard />, { user: authUser });

export { renderWithUserContext, renderWithAppContext, renderWithAllContexts, render };
