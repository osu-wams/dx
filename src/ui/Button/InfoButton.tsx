import React, { useEffect, useState, useContext } from 'react';
import VisuallyHidden from '@reach/visually-hidden';
import styled, { ThemeContext } from 'styled-components/macro';
import { fontSize } from 'src/theme';
import Icon from 'src/ui/Icon';
import MyDialog from 'src/ui/MyDialog';
import { CloseButton } from './index';
import { faInfoCircle } from '@fortawesome/pro-light-svg-icons';
import Button from './Button';
import { Event } from 'src/util/gaTracking';
import { InfoButtonState } from '@osu-wams/hooks/dist/api/infoButtons';
import { useRecoilValue } from 'recoil';
import { infoButtonState } from 'src/state/application';

const DialogHeader = styled.div`
  display: flex;
  flex-direction: row;
`;

const DialogClose = styled(CloseButton)`
  float: none;
  padding: 0;
`;

const DialogTitle = styled.h2`
  flex-grow: 2;
  font-size: ${fontSize['20']};
`;

const DialogContent = styled.div`
  font-size: ${fontSize['14']};
`;

const InfoButton = (props) => {
  const themeContext = useContext(ThemeContext);
  const infoButtonData = useRecoilValue(infoButtonState);
  const [dialogVisible, toggleDialog] = useState(false);
  const [currentButton, setButton] = useState<InfoButtonState | null>(null);

  useEffect(() => {
    if (Array.isArray(infoButtonData)) {
      const thisButton = infoButtonData.find((i) => i.id === props.infoButtonId);
      if (thisButton) {
        setButton(thisButton);
      }
    }
  }, [infoButtonData, props.infoButtonId]);

  return currentButton ? (
    <>
      <Button
        bg={themeContext.ui.button.info.background}
        data-testid={props.infoButtonId}
        onClick={(e: React.MouseEvent<HTMLElement>) => {
          toggleDialog(true);
          Event('info-button', currentButton.title);
        }}
      >
        <Icon icon={faInfoCircle} size="lg" color={themeContext.ui.button.info.icon.color} />
        <VisuallyHidden>Information about {currentButton.title}</VisuallyHidden>
      </Button>
      <MyDialog
        isOpen={dialogVisible}
        onDismiss={() => toggleDialog(false)}
        aria-labelledby="infobtn-title"
      >
        <DialogHeader>
          <DialogTitle id="infobtn-title">{currentButton.title}</DialogTitle>
          <DialogClose onClick={(e: React.MouseEvent<HTMLElement>) => toggleDialog(false)} />
        </DialogHeader>
        <DialogContent dangerouslySetInnerHTML={{ __html: currentButton.content }} />
      </MyDialog>
    </>
  ) : (
    <></>
  );
};

export default InfoButton;
