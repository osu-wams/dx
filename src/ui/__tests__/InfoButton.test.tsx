import React from 'react';
import { render } from 'src/util/test-utils';
import { InfoButton } from '../Button';
import { infoButtonState } from 'src/state/application';

it('does not render an info button because no id is supplied', () => {
  const { container } = render(<InfoButton />);

  expect(container.firstChild).toMatchSnapshot();
});

describe('with an application state provided', () => {
  const mockInitialState = jest.fn();

  beforeEach(() => {
    mockInitialState.mockReturnValue([
      {
        state: infoButtonState,
        value: [{ content: 'Info Button Content', id: 'valid-id', title: 'Info Button Title' }],
      },
    ]);
  });

  it('renders an info button ', () => {
    const { container } = render(<InfoButton infoButtonId="valid-id" />, {
      initialStates: mockInitialState(),
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('does not render an info button with an invalid id', () => {
    const { container } = render(<InfoButton infoButtonId="missing-id" />, {
      initialStates: mockInitialState(),
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
