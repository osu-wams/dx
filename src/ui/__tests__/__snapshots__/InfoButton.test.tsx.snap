// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render an info button because no id is supplied 1`] = `null`;

exports[`with an application state provided does not render an info button with an invalid id 1`] = `null`;

exports[`with an application state provided renders an info button  1`] = `
.c1 {
  background-color: transparent;
  color: #ffffff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: .6rem 1.4rem;
}

.c1 + .c0 {
  margin-left: 8px;
}

.c2 {
  color: #696361;
  background-color: transparent;
}

@media screen and (max-width:768px) {

}

@media (min-width:768px) {

}

<button
  class="c0 c1"
  data-testid="valid-id"
>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-info-circle fa-w-16 fa-lg c2"
    color="#696361"
    data-icon="info-circle"
    data-prefix="fal"
    focusable="false"
    role="img"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M256 40c118.621 0 216 96.075 216 216 0 119.291-96.61 216-216 216-119.244 0-216-96.562-216-216 0-119.203 96.602-216 216-216m0-32C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm-36 344h12V232h-12c-6.627 0-12-5.373-12-12v-8c0-6.627 5.373-12 12-12h48c6.627 0 12 5.373 12 12v140h12c6.627 0 12 5.373 12 12v8c0 6.627-5.373 12-12 12h-72c-6.627 0-12-5.373-12-12v-8c0-6.627 5.373-12 12-12zm36-240c-17.673 0-32 14.327-32 32s14.327 32 32 32 32-14.327 32-32-14.327-32-32-32z"
      fill="currentColor"
    />
  </svg>
  <span
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap; word-wrap: normal;"
  >
    Information about 
    Info Button Title
  </span>
</button>
`;
